README.md
## Features

Flutter屏幕适配方案 / Flutter screen adaptation solution

## Screenshot

| <img src="https://github.com/hwh97/screenshot_res/blob/master/flutter_screen_adapter/s1.png" width="340px" alt="screenshot"> | <img src="https://github.com/hwh97/screenshot_res/blob/master/flutter_screen_adapter/s1_landscape.png" width="340px" alt="screenshot"> |
|:----------------------------------------------------------------------------------------------------------------------------:|:--------------------------------------------------------------------------------------------------------------------------------------:|
| <img src="https://github.com/hwh97/screenshot_res/blob/master/flutter_screen_adapter/s2.png" width="340px" alt="screenshot"> | <img src="https://github.com/hwh97/screenshot_res/blob/master/flutter_screen_adapter/s2_landscape.png" width="340px" alt="screenshot"> |

## Usage

### Step1:
Import package
```
flutter_screen_adapter:
    git:
        url: git@github.com:hwh97/flutter_screen_adapter.git 
```
### Step2:
Add builder to MaterialApp
```
MaterialApp(
    builder: (context, child) => screenAdapterBuilder(
        child: child,
        designSize: const Size(375, 812),
        tabletDesignSize: const Size(768, 1024),
    ),
)
```

## Note
This plugin uses width >= 550 to get tablet or not. Feel free to pr if you have a better way.

## Thanks
https://juejin.cn/post/6844903822188363789

https://pub.flutter-io.cn/packages/responsive_framework

CHANGELOG.md
## 0.0.6
* add utils

## 0.0.5
* fix bug 

## 0.0.4
* replace WidgetBuilder with Widget

## 0.0.3
* fix values of padding,viewInsets,viewPadding
* update example

## 0.0.2
* replace WidgetsFlutterBinding with MediaQuery, which make it less invasive.
* add tablet adapter
* update the values of padding,viewInsets,viewPadding

## 0.0.1

* initial release.


test/flutter_screen_adapter_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_screen_adapter/flutter_screen_adapter.dart';

void main() {

}


lib/src/media_query_builder.dart
import 'dart:ui';

import 'package:flutter/widgets.dart';

Widget screenAdapterBuilder({
  required WidgetBuilder? builder,
  required Size designSize,
  Size? tabletDesignSize,
}) {
  return _MediaQueryBuilder(
    builder: builder!,
    designSize: designSize,
    tabletDesignSize: tabletDesignSize,
  );
}

/// utils
class ScreenAdapterUtil {
  ScreenAdapterUtil._();

  static _MediaQueryBuilderState _getState(BuildContext context) {
    final state = context.findAncestorStateOfType<_MediaQueryBuilderState>();
    assert(state != null);
    return state!;
  }

  static MediaQueryData originalMediaData(BuildContext context) {
    return _getState(context).originalMediaData;
  }

  static bool isTablet(BuildContext context) {
    return _getState(context).isTablet;
  }

  static double designWidth(BuildContext context) {
    return _getState(context).designWidth;
  }

  static Size scaleSize(BuildContext context) {
    return _getState(context).scaleSize;
  }

  static double scaleDevicePixelRatio(BuildContext context) {
    return _getState(context).devicePixelRatio;
  }

  static double scaleHeight(BuildContext context) {
    return _getState(context).scaleHeight;
  }

  static double devicePixelRatioRatio(BuildContext context) {
    return _getState(context).devicePixelRatioRatio;
  }

  static EdgeInsets padding(BuildContext context) {
    return _getState(context).padding;
  }

  static EdgeInsets viewInsets(BuildContext context) {
    return _getState(context).viewInsets;
  }

  static EdgeInsets viewPadding(BuildContext context) {
    return _getState(context).viewPadding;
  }
}

/// rewrite MediaQueryData
class _MediaQueryBuilder extends StatefulWidget {
  final WidgetBuilder builder;

  /// The size of the device screen in the design draft
  final Size designSize;

  /// The size of the tablat screen in the design draft
  final Size? tabletDesignSize;

  const _MediaQueryBuilder({
    Key? key,
    required this.builder,
    required this.designSize,
    this.tabletDesignSize,
  }) : super(key: key);

  @override
  State<_MediaQueryBuilder> createState() => _MediaQueryBuilderState();
}

class _MediaQueryBuilderState extends State<_MediaQueryBuilder> {
  /// Original data
  MediaQueryData get originalMediaData => MediaQuery.of(context);

  /// Get is tablet
  bool get isTablet {
    return MediaQuery.of(context).size.shortestSide >= 550;
  }

  /// Design width
  double get designWidth {
    final isLandscape =
        MediaQuery.of(context).orientation == Orientation.landscape;
    var designWidth =
        isLandscape ? widget.designSize.height : widget.designSize.width;

    if (widget.tabletDesignSize != null && isTablet) {
      designWidth = isLandscape
          ? widget.tabletDesignSize!.height
          : widget.tabletDesignSize!.width;
    }
    return designWidth;
  }

  /// Scale size
  Size get scaleSize => Size(designWidth, scaleHeight);

  /// New device pixel ratio
  double get devicePixelRatio => window.physicalSize.width / designWidth;

  /// New height by new [devicePixelRatio]
  double get scaleHeight => window.physicalSize.height / devicePixelRatio;

  /// devicePixelRatio Ratio
  double get devicePixelRatioRatio =>
      MediaQuery.of(context).devicePixelRatio / devicePixelRatio;

  /// new padding
  EdgeInsets get padding {
    final padding = MediaQuery.of(context).padding;
    return padding * devicePixelRatioRatio;
  }

  /// new viewInsets
  EdgeInsets get viewInsets {
    final viewInsets = MediaQuery.of(context).viewInsets;
    return viewInsets * devicePixelRatioRatio;
  }

  /// new viewPadding
  EdgeInsets get viewPadding {
    final viewPadding = MediaQuery.of(context).viewPadding;
    return viewPadding * devicePixelRatioRatio;
  }

  @override
  Widget build(BuildContext context) {
    return MediaQuery(
      child: FittedBox(
        fit: BoxFit.fitWidth,
        alignment: Alignment.topCenter,
        child: SizedBox(
          width: designWidth,
          height: scaleHeight,
          child: Builder(
            builder: widget.builder,
          ),
        ),
      ),
      data: MediaQuery.of(context).copyWith(
        size: scaleSize,
        devicePixelRatio: devicePixelRatio,
        padding: padding,
        viewInsets: viewInsets,
        viewPadding: viewPadding,
      ),
    );
  }
}


lib/src/screen_adapter_binding.dart
import 'dart:collection';
import 'dart:ui';
import 'package:flutter/gestures.dart';
import 'package:flutter/rendering.dart';
import 'package:flutter/widgets.dart';
import '/src/screen_util.dart';

/// rewrite runApp
runAdapterApp(Widget app) {
  ScreenAdapterBinding.ensureInitialized()!
    ..scheduleAttachRootWidget(app)
    ..scheduleWarmUpFrame();
}

/// rewrite WidgetsFlutterBinding
class ScreenAdapterBinding extends WidgetsFlutterBinding {
  final Queue<PointerEvent> _pendingPointerEvents = Queue<PointerEvent>();

  ScreenAdapterBinding();

  static WidgetsBinding? ensureInitialized() {
    if (WidgetsBinding.instance == null) {
      ScreenAdapterBinding();
    }
    return WidgetsBinding.instance;
  }

  @override
  void initInstances() {
    super.initInstances();
    window.onPointerDataPacket = _handlePointerDataPacket;
  }

  @override
  ViewConfiguration createViewConfiguration() {
    return ViewConfiguration(
      size: Size(
        ScreenUtil.designWidth,
        ScreenUtil.adapterHeight,
      ),
      devicePixelRatio: ScreenUtil.adapterPixelRatio,
    );
  }

  void _handlePointerDataPacket(PointerDataPacket packet) {
    // We convert pointer data to logical pixels so that e.g. the touch slop can be
    // defined in a device-independent manner.
    _pendingPointerEvents.addAll(PointerEventConverter.expand(
      packet.data,
      ScreenUtil.adapterPixelRatio,
    ));
    if (!locked) _flushPointerEventQueue();
  }

  void _flushPointerEventQueue() {
    assert(!locked);
    while (_pendingPointerEvents.isNotEmpty) {
      handlePointerEvent(_pendingPointerEvents.removeFirst());
    }
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter_screen_adapter/flutter_screen_adapter.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      debugShowCheckedModeBanner: false,
      builder: (context, child) => screenAdapterBuilder(
        builder: (context) => child!,
        designSize: const Size(375, 812),
        tabletDesignSize: const Size(768, 1024),
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Container(
              color: Colors.blueAccent,
              width: 100,
              height: 100,
              alignment: Alignment.center,
              child: const Text(
                '100 * 100',
                style: TextStyle(
                  fontSize: 16,
                  color: Colors.white,
                ),
              ),
            ),
            const SizedBox(height: 20),
            Text(
              '${MediaQuery.of(context)}',
              textAlign: TextAlign.center,
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
          child: const Icon(Icons.arrow_forward_rounded),
          onPressed: () {
            Navigator.push(context, MaterialPageRoute(
              builder: (context) {
                return const MyHomePage(title: "New Page");
              },
            ));
          }),
    );
  }
}
